SPLAT-1879: vSphere multi disk enhancement (#1709)
openshift/enhancements: enhancements/machine-api/vsphere-data-disk.md

- https://issues.redhat.com/browse/SPLAT-1880
see-also:
replaces:
superseded-by:
---

# vSphere Multi Disk

## Summary

This feature enhancement aims to allow admins the ability to configure additional disks that are not present in the vSphere VM templates (OVA) by enhancing the vsphere machine API and adding the capability to the cloning process.

## Motivation

As the use of Kubernetes clusters grows, admins are needing more and more improvements to the VMs themselves to make sure they run as smoothly as possible.  The number of cores and memory continue to increase for each machine and this is causing the amount of workloads to increase on each virtual machine.  This growth is now causing the base VM image to not provide enough storage for OS needs.  In some cases, users just increase the size of the primary disk using the existing configuration options for machines; however, this does not allows for all desired configuration choices.  Admins are now wanting the ability to add additional disks to these VMs for things such as etcd storage, image storage, container runtime and even swap.

### User Stories

> * As an OpenShift administrator, I want to be able to add additional disks to any of the vsphere VMs so that my nodes can have additional disks for me to use to assign special case storage such as etcd data, swap, container images, etc.

Reviewer: ```suggestion
* As an OpenShift administrator, I want to be able to add additional disks to any vsphere VMs which are acting as a node so nodes can have additional disks for me to use to assign special case storage such as etcd data, swap, container images, etc.
```

Author: Good catch

### Goals

- Enhance vSphere machine API to allow creating and attaching data disks to the VMs during cloning process.

> - Enhance OCP installer to supported adding data disks to control plane and compute machines.

Reviewer: ```suggestion
- Enhance OCP installer to support adding data disks to control plane and compute machines.
```

- Update OCP Cluster CAPI Operator / Cluster API vSphere to include new upstream API changes to support creating data disks.
- Keep upstream vSphere cluster API (CAPV) in sync with OCP Machine API (MAPI).

### Non-Goals

- Introduce API breaking changes.

> - Add ability to advance configure additions disks, such as, define controller type (IDE, scsi, etc) or unit number in the controller.

Reviewer: ```suggestion
- Add ability to perform advanced disk configuration, such as, defining controller type (IDE, scsi, etc) or unit number in the controller.
```

Author: I'll update that

#### Installer

The installer's install-config will be enhanced to allow the vSphere machine pools to define data disks.

```go
apiVersion: v1
baseDomain: openshift.example.net
featureSet: TechPreviewNoUpgrade
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform: 
    vsphere:
      zones:      
      - fd-1
      cpus: 4
      coresPerSocket: 2
      memoryMB: 8192
      osDisk:
        diskSizeGB: 60
      dataDisks:
      - diskSizeGiB: 10
        name: "container_images"
        provisioningMode: "Thin"
      - diskSizeGiB: 20
        name: "log files"
        provisioningMode: "Thick"
      - diskSizeGiB: 30
        name: "swap"
        provisioningMode: "EagerlyZeroed"
  replicas: 0
controlPlane:
  architecture: amd64
  hyperthreading: Enabled
  name: master
  platform:
    vsphere: 
      zones:
      - fd-1
      cpus: 8 
      coresPerSocket: 2
      memoryMB: 16384
      osDisk:
        diskSizeGB: 100
      dataDisks:
      - diskSizeGiB: 10
        name: "etcd"
        provisioningMode: "Thick"
      - diskSizeGiB: 20
        name: "container-images"
        provisioningMode: "Thin"
  replicas: 3
...
```

> In the example above, we have added a new field `dataDisks` that is used to define each of the disks to create for the virtual machine.  The `name` field is used to identify the disk.  This is used primarily for debug and for identifying the disk definition.  The `diskSizeGiB` field specifies how large the disk needs to be. The `provisioningMode` field specifies how the disk should be provisioned.  This field supports the values of "Thin", "Thick", and "EagerlyZeroed"

Reviewer: ```suggestion
In the example above, we have added a new field `dataDisks` that is used to define each of the disks to create for the virtual machine.  The `name` field is used to identify the disk.  This is used primarily for debugging and for identifying the disk definition.  The `diskSizeGiB` field specifies how large the disk needs to be. The `provisioningMode` field specifies how the disk should be provisioned.  This field supports the values of "Thin", "Thick", and "EagerlyZeroed"
```

    machines_v1beta1_machine_openshift_io:
      failureDomains:
        platform: VSphere
        vsphere:
        - name: fd-1
      metadata:
        labels:
          machine.openshift.io/cluster-api-cluster: ngirard-multi-rtmvw
          machine.openshift.io/cluster-api-machine-role: master
          machine.openshift.io/cluster-api-machine-type: master
      spec:
        lifecycleHooks: {}
        metadata: {}
        providerSpec:
          value:
            apiVersion: machine.openshift.io/v1beta1
            credentialsSecret:
              name: vsphere-cloud-credentials
            dataDisks:
            - diskSizeGiB: 10
              name: "etcd"
              provisioningMode: "Thick"
            - diskSizeGiB: 20
              name: "container-images"
              provisioningMode: "Thin"
            diskGiB: 100
            kind: VSphereMachineProviderSpec
            memoryMiB: 16384
            metadata:
              creationTimestamp: null
            network:
              devices: null
            numCPUs: 8
            numCoresPerSocket: 2
            snapshot: ""
            template: ""
            userDataSecret:
              name: master-user-data

>             workspace: {}

Reviewer: have we done any testing with multiple failure domains and additional data disks? We have some webhook/validation logic that I don't think would be impacted by this, but I just want to make sure.

Author: I have tested creating vms with multiple disks in one of the FDs in a multi FD cluster.  I can run a quick test where i spin up several across all FDs to validate this. These disks are created as part of the vm clone process.  I'd expect the webhook not to care much other than validating the vlaues of each disk config, but I'll rerun and confirm.

Author: Yep, it is all looking good.
